hypervisor/vm.go

```diff
@@ -635,30 +640,20 @@ func (vm *Vm) Pause(pause bool) error {
 		command = "Unpause"
 	}
 
-	err := vm.GenericOperation(command, func(ctx *VmContext, result chan<- error) {
-		oldPauseState = ctx.PauseState
-		if ctx.PauseState == PauseStateBusy {
-			result <- fmt.Errorf("%s fails: earlier Pause or Unpause operation has not finished", command)
-			return
-		} else if ctx.PauseState == pauseState {
-			result <- nil
-			return
-		}
-		ctx.PauseState = PauseStateBusy
+	var err error
+	ctx.pauseLock.Lock()
+	oldPauseState = ctx.PauseState
+	if ctx.PauseState != pauseState {
 		/* FIXME: only support pause whole vm now */
-		ctx.DCtx.Pause(ctx, pause, result)
-	}, StateRunning)
-
-	if oldPauseState == pauseState {
-		return nil
+		err = ctx.DCtx.Pause(ctx, pause)
 	}
 	if err != nil {
-		pauseState = oldPauseState // recover the state
+		glog.Errorf("%s sandbox failed, recover the state", command)
+		ctx.PauseState = oldPauseState // recover the state
```

Comment:
the above line and oldPauseState should be removed, ctx.PauseState has not been changed.

but you forgot to set ctx.PauseState after success.